lib/ReorderBasicBlocksPass.cpp

```diff
@@ -41,21 +54,94 @@ namespace clspv {
 FunctionPass *createReorderBasicBlocksPass() {
   return new ReorderBasicBlocksPass();
 }
+} // namespace clspv
+
+void ReorderBasicBlocksPass::StructuredOrder(BasicBlock *block,
+                                             const LoopInfo &LI,
+                                             std::deque<BasicBlock *> *order,
+                                             DenseSet<BasicBlock *> *visited) {
+  if (!visited->insert(block).second)
+    return;
+
+  // Identify the merge and continue blocks for special treatment.
+  const auto *terminator = dyn_cast<TerminatorInst>(block->getTerminator());
+  BasicBlock *continue_block = nullptr;
+  BasicBlock *merge_block = nullptr;
+  if (LI.isLoopHeader(block)) {
+    Loop *loop = LI.getLoopFor(block);
+    merge_block = loop->getExitBlock();
+
+    if (loop->isLoopLatch(block)) {
+      continue_block = block;
+    } else {
+      DominatorTree &DT = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
```

Comment:
// This "else" case is for when the loop has several blocks.

// The continue block satisfies the following conditions:
// 1. Is dominated by the header. (That's true by construction at this point.)
// 2. Dominates the latch block.
// We can assume the loop has multiple blocks, since that single block loop case was handled separately.  
// By construction, the header always dominate the latch block, so we have to exclude that case specifically.